app/src/main/java/org/mozilla/firefox/vpn/apptunneling/ui/ExpandableAdapter.kt

```diff
@@ -0,0 +1,161 @@
+            }
+            else -> {
+                (holder as AppGroupViewHolder).bind(items[position] as AppGroup)
+            }
+        }
+    }
+
+    private fun onAppGroupClicked(holder: AppGroupViewHolder) {
+        val appGroupItem = items[holder.adapterPosition] as AppGroup
+        val startPosition = holder.adapterPosition + 1
+        val count = appGroupItem.appItems.size
+
+        if (appGroupItem.isExpanded) {
+            items.removeAll(appGroupItem.appItems)
+            notifyItemRangeRemoved(startPosition, count)
+            (items[holder.adapterPosition] as AppGroup).isExpanded = false
+        } else {
+            items.addAll(startPosition, appGroupItem.appItems)
+            notifyItemRangeInserted(startPosition, count)
+            (items[holder.adapterPosition] as AppGroup).isExpanded = true
+        }
+        notifyItemChanged(holder.adapterPosition)
+    }
+
+    private fun onAppGroupActionClicked(holder: AppGroupViewHolder) {
+        val appGroupItem = items[holder.adapterPosition] as AppGroup
+        val packageNameSet = appGroupItem.appItems.map { it.applicationInfo.packageName }.toSet()
+
+        if (appGroupItem.type == AppGroupType.UNPROTECTED) {
+            expandableItemCallback.onProtectAllClicked(packageNameSet)
+        } else {
+            expandableItemCallback.onUnprotectAllClicked(packageNameSet)
+        }
+    }
+
+    private fun onAppGroupActionClicked(holder: AppItemViewHolder) {
+        val appItem = items[holder.adapterPosition] as AppItem
+        val packageName = appItem.applicationInfo.packageName
+
+        if (appItem.type == AppGroupType.UNPROTECTED) {
+            expandableItemCallback.onUnprotectedAppChecked(packageName)
+        } else {
+            expandableItemCallback.onProtectedAppChecked(packageName)
+        }
+    }
+
+    private fun AppTunnelingUiModel.toExpandableList(): MutableList<ExpandableItem> {
+        val list = mutableListOf<ExpandableItem>()
+        val (exclude, include) = packageList.partition { excludeList.contains(it.packageName) }
+        val excludeAppItems = exclude.map { AppItem(AppGroupType.UNPROTECTED, it) }
+        val includeAppItems = include.map { AppItem(AppGroupType.PROTECTED, it) }
+
+        list.add(AppGroup(AppGroupType.UNPROTECTED, excludeAppItems))
+        list.addAll(excludeAppItems)
+        list.add(AppGroup(AppGroupType.PROTECTED, includeAppItems))
+        list.addAll(includeAppItems)
+        return list
+    }
+
+    private fun getDiffResult(oldList: List<ExpandableItem>, newList: List<ExpandableItem>): DiffUtil.DiffResult {
+
+        return DiffUtil.calculateDiff(object : DiffUtil.Callback() {
+
+            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
+                val newItem = newList[newItemPosition]
+                val oldItem = oldList[oldItemPosition]
+
+                return newItem.itemType == oldItem.itemType
```

Comment:
areItemsTheSame should tell whether two items (one from the old list, and one from the new list) are the same, for example, calculator and clock are two different apps, thus they are not the same even if their view types are the same.